### DISCUSSION TITLE
Today I spent 20€ on Claude building a GNURadio port for micropython

### DISCUSSION THREAD

**Discussion by jgpeiro** [micropython/micropython#14386] [Ideas]
The project took me about two hours and the reuslts contains a decent set of blocks, with unittests and examples. The whole code is implemented in python without any dependence. (even some np's still needs to be removed).
The tests pass without mayor issues and MagicMock, used for testing, was also generated.
Common peripherials, like UARTs, SPIs, ADC, PWM... are supported. It can use files or udp connexions too.
The message system to communicate different blocks is really simple.
The full conversation is available at prompt.txt
(The readme of the project is also autogenerated.)

[https://github.com/jgpeiro/ugnuradio](https://github.com/jgpeiro/ugnuradio)

**Comment by andypiper** [comment]
Looks like your README needs editing to add contact information, and the license link is broken.

Have you tested this? What hardware have you run it on? What version of MicroPython does it work with?

**Comment by peterhinch** [comment]
@jgpeiro re README.md the clone instruction should read `git clone https://github.com/jgpeiro/ugnuradio.git`.

The README also includes the line
> For a complete list of supported blocks and their usage, please refer to the documentation.

Which documentation? There is none in the repository other than the README.

Looking at the `demo_analog.py`, the following line seems to reference a nonexistent module:
```py
from device_source import DeviceSource
```
The other imports will fail because the modules are in a subdirectory `ugr`. Have these demos ever actually been run?

The demo itself references various pin numbers: clearly these can only be appropriate on a particular platform. Which one?

More documentation is needed if anyone is to actually use this. Some idea of performance would also be useful. What is the bandwidth of the analog demo?

**Reply by jgpeiro** [reply]
Sorry, I didn't express myself well. The entire project is generated with AI, including the readme (to which I just added a note) are generated and are not correct.
This project took me only two hours, and its only a proof of concept.

The real status of the code is that some basic functionality is already possible and the multiple
blocks available looks promising.

The unittests, also generated, works as expected too.

Im using a Pico-W with micropython 1.22 I think.

The foto of the readme (and the readme itself) does not reflect the actual code functionality.

I tested few of the blocks with this code (from the file demo_gnuradio.py) and it worked fine.
```python
import time
from top_block import TopBlock
from sig_source import SigSource
from throttle import Throttle
from file_sink import FileSink

class MyGraph(TopBlock):
    def __init__(self):
        super().__init__()

        ##################################################
        # Variables
        ##################################################
        self.samp_rate = 100

        ##################################################
        # Blocks
        ##################################################
        self.analog_sig_source_x_0 = SigSource(self.samp_rate, "GR_COS_WAVE", 10, 1000, 0)
        self.blocks_throttle_0 = Throttle(2, self.samp_rate)
        #self.blocks_udp_sink_0 = UDPSink(2, '192.168.1.103', 1234, 1024, True)
        self.blocks_file_sink_0 = FileSink(2, "output.txt")
        
        ##################################################
        # Add blocks (required on this port)
        ##################################################
        self.add_block(self.analog_sig_source_x_0)
        self.add_block(self.blocks_throttle_0)
        self.add_block(self.blocks_file_sink_0)
        
        ##################################################
        # Connections
        ##################################################
        self.connect(self.analog_sig_source_x_0, self.blocks_throttle_0)
        self.connect(self.blocks_throttle_0, self.blocks_file_sink_0)

    def get_samp_rate(self):
        return self.samp_rate

    def set_samp_rate(self, samp_rate):
        self.samp_rate = samp_rate
        self.analog_sig_source_x_0.sampling_freq = self.samp_rate
        self.blocks_throttle_0.sample_rate = self.samp_rate


def main():
    tb = MyGraph()
    tb.start()
    time.sleep_ms(10000)
    tb.stop()
    tb.wait()

if __name__ == '__main__':
    main()
```
